### PR TITLE
Add base class and two implementations of record filters

### PR DESCRIPTION
**Pull request by cmcantalupo**
- Add doxygen comments to filter classes
- Add unit test for EpochRecordFilter
- Add unit tests for the ProxyEpochRecordFilter

### CODE REVIEW

File: test/ProxyEpochRecordFilterTest.cpp [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+        EXPECT_EQ(event, result[0].event);
+        EXPECT_EQ(0ULL, result[0].signal);
+    }
+}
+
+TEST_F(ProxyEpochRecordFilterTest, filter_out)
+{
+    ApplicationSampler::m_record_s record {};
+    std::vector<ApplicationSampler::m_record_s> result;
+    geopm::ProxyEpochRecordFilter perf(0XAULL, 1, 0);
```
Comment (cmcantalupo) [inline test/ProxyEpochRecordFilterTest.cpp]: 0xAULL for consistency.

File: src/EpochRecordFilter.hpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+            EpochRecordFilter() = default;
+            /// @brief Default destructor.
+            virtual ~EpochRecordFilter() = default;
+            /// @brief The M_EVENT_EPOCH and M_EVENT_HINT event types
+            ///        are retained, all other records are removed
+            ///        from the stream.
+            ///
+            /// @param record The update value to be filtered.
+            ///
+            /// @return Vector of lenth one containing input value if
```
Comment (dianarg) [inline src/EpochRecordFilter.hpp]: length

File: src/ProxyEpochRecordFilter.cpp [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+namespace geopm
+{
+    ProxyEpochRecordFilter::ProxyEpochRecordFilter(uint64_t region_hash,
+                                                   int calls_per_epoch,
+                                                   int startup_count)
+        : m_proxy_hash(region_hash)
+        , m_num_per_epoch(calls_per_epoch)
+        , m_count(-startup_count)
+    {
+        if (m_proxy_hash > 1ULL <<32) {
```
Comment (dianarg) [inline src/ProxyEpochRecordFilter.cpp]: maybe a comment or name for this magic number.  hopefully we have some documentation about this somewhere

File: src/ProxyEpochRecordFilter.cpp [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    std::vector<ApplicationSampler::m_record_s> ProxyEpochRecordFilter::filter(const ApplicationSampler::m_record_s &record)
+    {
+        std::vector<ApplicationSampler::m_record_s> result;
+        if (record.event == ApplicationSampler::M_EVENT_REGION_ENTRY &&
+            record.signal == m_proxy_hash) {
+            if (m_count >= 0 &&
+                m_count % m_num_per_epoch == 0) {
+                ApplicationSampler::m_record_s epoch_event(record);
```
Comment (dianarg) [inline src/ProxyEpochRecordFilter.cpp]: kind of a nit, but assignment would be nicer than copy constructor

File: src/EpochRecordFilter.hpp [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        public:
+            /// @brief Default constructor.
+            EpochRecordFilter() = default;
+            /// @brief Default destructor.
+            virtual ~EpochRecordFilter() = default;
+            /// @brief The M_EVENT_EPOCH and M_EVENT_HINT event types
+            ///        are retained, all other records are removed
+            ///        from the stream.
+            ///
+            /// @param record The update value to be filtered.
```
Comment (dianarg) [inline src/EpochRecordFilter.hpp]: I think these should be `@param [in] name`.  more below.

File: src/ProxyEpochRecordFilter.hpp [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    /// region prior to the beginning of the outer loop.  The filter
+    /// expects to be provided records collected from a single
+    /// process.
+    class ProxyEpochRecordFilter : public RecordFilter
+    {
+        public:
+            /// @brief Constructor for a process specific proxy-region
+            ///        EpochIOGroup record filter.
+            ///
+            /// @param region_hash The hash for the region that will
```
Comment (dianarg) [inline src/ProxyEpochRecordFilter.hpp]: [in]

File: test/ProxyEpochRecordFilterTest.cpp [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
+TEST_F(ProxyEpochRecordFilterTest, skip_two_off_one)
+{
+    uint64_t hash = 0xAULL;
+    ApplicationSampler::m_record_s record {0.0,
+                                           0,
+                                           ApplicationSampler::M_EVENT_REGION_ENTRY,
+                                           hash};
+    geopm::ProxyEpochRecordFilter perf(hash, 3, 1);
+    perf.filter(record);
```
Comment (dianarg) [inline test/ProxyEpochRecordFilterTest.cpp]: assert size == 0 here ?

**Review by dianarg — CHANGES_REQUESTED** [review]

**Review by dianarg — APPROVED** [review]
LGTM

**Pull request closed without merging**